### PR TITLE
rgw: list_bucket versions return NextVersionIdMarker = "null" if next_marker.instance is empty

### PR DESCRIPTION
**Pull request by shashalu**
Signed-off-by: Shasha Lu <lu.shasha@eisoo.com>

### CODE REVIEW

**Comment by shashalu** [comment]
@cbodley please take a look, thx.

File: src/rgw/rgw_rest_s3.cc [hunk 1]
```diff
@@ -684,7 +684,10 @@ void RGWListBucket_ObjStore_S3::send_versioned_response()
   s->formatter->dump_string("VersionIdMarker", marker.instance);
   if (is_truncated && !next_marker.empty()) {
     s->formatter->dump_string("NextKeyMarker", next_marker.name);
-    s->formatter->dump_string("NextVersionIdMarker", next_marker.instance);
+    if (next_marker.instance.empty())
+      s->formatter->dump_string("NextVersionIdMarker", "null");
```
Comment (ZVampirEM77) [inline src/rgw/rgw_rest_s3.cc]: I think keep it as a empty string is OK.

Think about the using case that someone get "NextVersionIdMarker" field, then decide whether use this field or not by judging "NextVersionIdMarker.empty()".

Comment (joscollin) [reply src/rgw/rgw_rest_s3.cc]: @ZVampirEM77 That's right. It will never be empty now.

Comment (ZVampirEM77) [reply src/rgw/rgw_rest_s3.cc]: @shashalu @joscollin Maybe it is my bad. I find S3 really returns "null" when some field is empty.

eg:
When I list object versions, if the bucket versioning function is disable, the response msg likes following:
{u'Name': 'XXX', 'ResponseMetadata': {'HTTPStatusCode': 200, 'RetryAttempts': 0, 
...
u'Versions': [{
...
u'VersionId': 'null', 
...
, u'Prefix': '', u'KeyMarker': '', u'IsTruncated': False, u'VersionIdMarker': ''}

Yeah, the "VersionId" is set to "null".

@shashalu  So could you try to confirm what will S3 return if the "NextVersionIdMarker" is empty. In my mind, the response msg of S3 will not include the "NextVersionIdMarker" field when "NextVersionIdMarker" is empty. Is it right?

Comment (shashalu) [reply src/rgw/rgw_rest_s3.cc]: I don't know what will S3 return.
I modify this just want to solve the following case.

There are three objects in a bucket.
objecta1 VersionId=1
objecta1 VersionId=null
objecta2 VersionId=1

list_bucket_versions with max-keys=2, original response will be "NextKeyMarker = objecta1; NextVersionIdMaker is empty"
Then use the NextKeyMarker and NextVersionIdMarker to list the remain objects in this bucket.
The response will be start with objecta1,VersionId=1. Actually, the expected response is start with objecta2,VersionId=1.

If using this "NextKeyMarker = objecta1; NextVersionIdMaker=null" to list the remain objects, the response is correct.

Comment (ZVampirEM77) [reply src/rgw/rgw_rest_s3.cc]: @shashalu For keeping compatible with S3, I think we should have a consistent response msg here.

**Comment by joke-lee** [comment]
i think this is a important bug fix,

i create bucket test2 and upload 1.txt 2.txt 3.txt ,then enable bucket versioning, and upload 2.txt twice,

so this 

```
1.txt null
2.txt rvxlH5ZD63Dxk87aa05QZAs3vPgK9IQT
2.txt 9udc3VRe1qz5tGWD9y3NGNrCOnjnDetU
2.txt null
3.txt null
```
and then  request

```
GET /test2/?versions&key-marker=1.txt&version-id-marker=null&max-keys=3
```

in the old version code

```
<?xml version='1.0' encoding='UTF-8'?>
<ListVersionsResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/">
  <Name>testversionsbeijing1</Name>
  <Prefix/>
  <KeyMarker>1.txt</KeyMarker>
  <VersionIdMarker>null</VersionIdMarker>
  <NextKeyMarker>2.txt</NextKeyMarker>
  <NextVersionIdMarker/>
  <MaxKeys>3</MaxKeys>
  <IsTruncated>true</IsTruncated>
  <EncodingType>url</EncodingType>
```

but in aws

```
<?xml version='1.0' encoding='UTF-8'?>
<ListVersionsResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/">
  <Name>testversionsbeijing1</Name>
  <Prefix/>
  <KeyMarker>1.txt</KeyMarker>
  <VersionIdMarker>null</VersionIdMarker>
  <NextKeyMarker>2.txt</NextKeyMarker>
  <NextVersionIdMarker>null</NextVersionIdMarker>
  <MaxKeys>3</MaxKeys>
  <IsTruncated>true</IsTruncated>
  <EncodingType>url</EncodingType>
```

some code like this
```
        ListVersionsRequest listVersionsRequest = new ListVersionsRequest();
        listVersionsRequest.setBucketName(bucketname);
        listVersionsRequest.withKeyMarker("1.txt");
        listVersionsRequest.withVersionIdMarker("null");
        listVersionsRequest.withMaxResults(3);
        VersionListing versionListing = s3.listVersions(listVersionsRequest);
        List<String> versionIds = new ArrayList<String>();
        int maxnum = 1;
        do {
            for (S3VersionSummary s3VersionSummary : versionListing.getVersionSummaries()) {
                System.out.println(s3VersionSummary.getKey() + " versionid= " + s3VersionSummary.getVersionId());
            }
            versionListing = s3.listNextBatchOfVersions(versionListing);
        }
        while (versionListing.isTruncated());
```

will loop forever !!!

hi @cbodley do you mind help review this?

**Comment by joke-lee** [comment]
unstale me, my man

**Comment by mattbenjamin** [comment]
apparently, this needs a teuthology run

**Comment by cbodley** [comment]
http://pulpito.ceph.com/cbodley-2019-04-23_20:35:16-rgw-wip-cbodley-testing-distro-basic-smithi/

lots of valgrind issues from http://tracker.ceph.com/issues/39449

**Pull request merged**